Unreproducible Package:  libsForQt5.qtwebengine.out

### Nixpkgs Revision

698214a3

### Introduction

Building **qtwebengine** multiple times does not yield bit-by-bit identical
results, complicating the detection of Continuous Integration (CI) breaches. For
more information on this issue, visit [reproducible-builds.org](https://reproducible-builds.org/).

Fixing bit-by-bit reproducibility also has additional advantages, such as
avoiding hard-to-reproduce bugs, making content-addressed storage more effective
and reducing rebuilds in such systems.


### Steps to reproduce

### 1. Build the package

This step will build the package. Specific arguments are passed to the command
to keep the build artifacts so we can compare them in case of differences.

Execute the following command:

```
nix-build '<nixpkgs>' -A qt5.qtwebengine.out && nix-build '<nixpkgs>' -A qt5.qtwebengine.out --check --keep-failed
```

Or using the new command line style:

```
nix build nixpkgs#qt5.qtwebengine.out && nix build nixpkgs#qt5.qtwebengine.out --rebuild --keep-failed
```

### 2. Compare the build artifacts

If the previous command completes successfully, no differences were found and
there's nothing to do, builds are reproducible.
If it terminates with the error message `error: derivation '<X>' may not be
deterministic: output '<Y>' differs from '<Z>'`, use `diffoscope` to investigate
the discrepancies between the two build outputs. You may need to add the
`--exclude-directory-metadata recursive` option to ignore files and directories
metadata (*e.g. timestamp*) differences.

```
nix run nixpkgs#diffoscopeMinimal -- --exclude-directory-metadata recursive <Y> <Z>
```

### 3. Examine the build log

To examine the build log, use:

```
nix-store --read-log $(nix-instantiate '<nixpkgs>' -A qt5.qtwebengine.out)
```

Or with the new command line style:

```
nix log $(nix path-info --derivation nixpkgs#qt5.qtwebengine.out)
```


### Diffoscope log

https://reproducible.nixos.org/diffoscope/qtwebengine-698214a3

### Notify maintainers

@matthewbauer 

---

**Note for maintainers:** Please tag this issue in your pull request description. (i.e. `Resolves #ISSUE`.)


### I assert that this issue is relevant for Nixpkgs

- [x] I assert that this is not a [duplicate of any known issue](https://github.com/NixOS/nixpkgs/issues?q=is%3Aopen+is%3Aissue+label%3A%226.topic%3A+reproducible+builds%22).
- [x] I assert that I have read the [NixOS Code of Conduct](https://github.com/NixOS/.github/blob/master/CODE_OF_CONDUCT.md) and agree to abide by it.

### Is this issue important to you?

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

> Ouch! Was that building just qtwebengine, or did it want to rebuild other artifacts as well? Was it really the 'build' step that was so slow or was downloading the dependencies a significant part of it?

I lost the previous logs, not sure what step took the longest but now I'm running again and it looks like it's the buildPhase that's taking.
```sh
[1597/24855] ACTION //base:build_date(/build/source/src/toolchain:target)chain:target)Kvent.pbzero.o.oKKo.oro.o[Ke/src/toolchain:target)
/build/source/src/core/release/../../3rdparty/chromium/build/write_build_date_header.py:19: DeprecationWarning: datetime.datetime.utcfromtimestamp() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.fromtimestamp(timestamp, datetime.UTC).
  date = datetime.datetime.utcfromtimestamp(int(args.timestamp))
[1793/24855] CXX obj/third_party/icu/bundled_icui18n/decNumber.oion-string-to-double.octors.o[Kpl.o[Ke-copy.stamp
../../3rdparty/chromium/third_party/icu/source/i18n/decNumber.cpp: In function 'decNumber* uprv_decNumberLog10_67(decNumber*, const decNumber*, decContext*)':
../../3rdparty/chromium/third_party/icu/source/i18n/decNumber.cpp:1528:14: warning: writing 1 byte into a region of size 0 [-Wstringop-overflow=]
 1528 |     w->lsu[1]=1; w->lsu[0]=0;           /* ..  */
      |     ~~~~~~~~~^~
In file included from ../../3rdparty/chromium/third_party/icu/source/i18n/decNumber.cpp:184:
../../3rdparty/chromium/third_party/icu/source/i18n/decNumber.h:89:19: note: at offset 1 into destination object 'decNumber::lsu' of size 1
   89 |     decNumberUnit lsu[DECNUMUNITS];
      |                   ^~~
../../3rdparty/chromium/third_party/icu/source/i18n/decNumber.cpp:1528:14: warning: writing 1 byte into a region of size 0 [-Wstringop-overflow=]
 1528 |     w->lsu[1]=1; w->lsu[0]=0;           /* ..  */
      |     ~~~~~~~~~^~
../../3rdparty/chromium/third_party/icu/source/i18n/decNumber.h:89:19: note: at offset 1 into destination object 'decNumber::lsu' of size 1
   89 |     decNumberUnit lsu[DECNUMUNITS];
```
Not sure what's happening here but it looks like it's building Chromium and the build system is churning through thousands of files. Currently in 1793/24855

--- a/pkgs/development/libraries/qt-6/modules/qtwebengine/default.nix
+++ b/pkgs/development/libraries/qt-6/modules/qtwebengine/default.nix
@@ -113,6 +113,8 @@ qtModule {
 
     # Fix build with Pipewire 1.4
     ./pipewire-1.4.patch
+    # Reproducibility QTBUG-136068
+    ./gn-object-sorted.patch
   ];
 
   postPatch =

--- a/pkgs/development/libraries/qt-6/modules/qtwebengine/gn-object-sorted.patch
+++ b/pkgs/development/libraries/qt-6/modules/qtwebengine/gn-object-sorted.patch
@@ -1,0 +1,32 @@
+diff --git a/gn/src/gn/rsp_target_writer.cc b/gn/src/gn/rsp_target_writer.cc
+index 6c1c687e99d..097b84b317e 100644
+--- a/src/3rdparty/gn/src/gn/rsp_target_writer.cc
++++ b/src/3rdparty/gn/src/gn/rsp_target_writer.cc
+@@ -117,8 +117,7 @@ void RspTargetWriter::Run() {
+       PathOutput path_output(settings->build_settings()->build_dir(),
+                              settings->build_settings()->root_path_utf8(),
+                              ESCAPE_NONE);
+-      std::vector<SourceFile> object_files;
+-      object_files.reserve(target_->sources().size());
++      std::set<SourceFile> object_files;
+ 
+       for (const auto& source : target_->sources()) {
+         const char* tool_type = nullptr;
+@@ -128,7 +127,7 @@ void RspTargetWriter::Run() {
+             other_files.push_back(source);
+           continue;  // No output for this source.
+         }
+-        object_files.push_back(
++        object_files.insert(
+             tool_outputs[0].AsSourceFile(settings->build_settings()));
+       }
+       if (target_->config_values().has_precompiled_headers()) {
+@@ -137,7 +136,7 @@ void RspTargetWriter::Run() {
+         if (tool_cxx && tool_cxx->precompiled_header_type() == CTool::PCH_MSVC) {
+           GetPCHOutputFiles(target_, CTool::kCToolCxx, &tool_outputs);
+           if (!tool_outputs.empty())
+-            object_files.push_back(
++            object_files.insert(
+                 tool_outputs[0].AsSourceFile(settings->build_settings()));
+         }
+       }

